### PR TITLE
feat: [CI-12042]: Add OIDC auth support

### PR DESCRIPTION
**Pull request by abhay084**
Please read [CONTRIBUTING.md](CONTRIBUTING.md) for details on our [code of conduct](CODE_OF_CONDUCT.md), and the process for submitting pull requests to us.

Fixes #

## Proposed Changes

-
-
-

Added the Support for OIDC-based Authentication for GCS.

I tested the change with the OIDC connector with my new change and current flow.

Screenshot 2024-04-23 at 3 37 03 PM
Screenshot 2024-04-23 at 3 39 21 PM
Screenshot 2024-04-23 at 3 39 37 PM

### Checklist

- [ ] Read the [**CONTRIBUTING**](CONTRIBUTING.md) document.
- [ ] Read the [**CODE OF CONDUCT**](CODE_OF_CONDUCT.md) document.
- [ ] Add tests to cover changes.
- [ ] Ensure your code follows the code style of this project.
- [ ] Ensure CI and all other PR checks are green OR
    - [ ] Code compiles correctly.
    - [ ] Created tests which fail without the change (if possible).
    - [ ] All new and existing tests passed.
- [ ] Add your changes to `Unreleased` section of [CHANGELOG](CHANGELOG.md).
- [ ] Improve and update the [README](README.md) (if necessary).
- [ ] Ensure [documentation](./DOCS.md) is up-to-date. The same file will be updated in [plugin index](https://github.com/drone/drone-plugin-index/blob/master/content/meltwater/drone-cache/index.md) when your PR is accepted, so it will be available for end-users at http://plugins.drone.io.

### CODE REVIEW

File: internal/gcp/tokenutil.go [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+	"context"
+	"fmt"
+
+	"golang.org/x/oauth2"
+	"google.golang.org/api/iamcredentials/v1"
+	"google.golang.org/api/option"
+	"google.golang.org/api/sts/v1"
+)
+
+type staticTokenSource struct {
```
Comment (Sapa96) [inline internal/gcp/tokenutil.go:13]: Do we need a separate structure for this? Can we use `oauth2.Token` directly?

File: internal/gcp/tokenutil.go [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+	return s.token, nil
+}
+
+func GetFederalToken(idToken, projectNumber, poolId, providerId string) (string, error) {
+	ctx := context.Background()
+	stsService, err := sts.NewService(ctx, option.WithoutAuthentication())
+	if err != nil {
+		return "", err
+	}
+	audience := fmt.Sprintf("//iam.googleapis.com/projects/%s/locations/global/workloadIdentityPools/%s/providers/%s", projectNumber, poolId, providerId)
```
Comment (Sapa96) [inline internal/gcp/tokenutil.go:27]: can we define url as a constant ?

File: internal/gcp/tokenutil.go [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+		token: &oauth2.Token{AccessToken: federatedToken},
+	}
+	service, err := iamcredentials.NewService(ctx, option.WithTokenSource(tokenSource))
+	if err != nil {
+		return "", err
+	}
+
+	name := "projects/-/serviceAccounts/" + serviceAccountEmail
+	rb := &iamcredentials.GenerateAccessTokenRequest{
+		Scope: []string{"https://www.googleapis.com/auth/cloud-platform"},
```
Comment (Sapa96) [inline internal/gcp/tokenutil.go:56]: same here

**Review by Sapa96 — APPROVED** [review]
LGTM

File: internal/gcp/tokenutil.go [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+	token := &oauth2.Token{AccessToken: federatedToken}
+	service, err := iamcredentials.NewService(ctx, option.WithTokenSource(oauth2.StaticTokenSource(token)))
+	if err != nil {
+		return "", err
+	}
+
+	name := "projects/-/serviceAccounts/" + serviceAccountEmail
+	// rb (request body) specifies parameters for generating an access token.
+	rb := &iamcredentials.GenerateAccessTokenRequest{
+		Scope: []string{"https://www.googleapis.com/auth/cloud-platform"},
```
Comment (Sapa96) [inline internal/gcp/tokenutil.go:55]: use the constant defined **scopeURL** here

**Review by raghavharness — APPROVED** [review]

**Review by harshjain12 — APPROVED** [review]

**Pull request merged**